toolkit/offchain/src/d_param.rs

```diff
@@ -74,109 +75,135 @@ fn get_current_d_parameter(
+	let spend_ex_units = get_validator_budgets(evaluate_response)?;
 
-	let tx = update_d_param_tx(validator, policy, d_parameter, current_utxo, &ctx, spend_ex_units)?;
+	let tx = update_d_param_tx(validator, policy, d_parameter, current_utxo, &ctx, spend_ex_units, genesis_utxo, client).await?;
 	let signed_tx = ctx.sign(&tx).to_bytes();
 	let res = client.submit_transaction(&signed_tx).await.map_err(|e| {
 		anyhow!(
 			"Submit D-parameter update transaction request failed: {}, bytes: {}",
 			e,
-			hex::encode(tx.to_bytes())
+			hex::encode(signed_tx)
 		)
 	})?;
 	log::info!("Update D-parameter transaction submitted: {}", hex::encode(res.transaction.id));
 	// TODO: await for the transaction output to be visible in validator wallet
 	Ok(McTxHash(res.transaction.id))
 }
 
-fn mint_d_param_token_tx(
+async fn mint_d_param_token_tx<C: QueryLedgerState + Transactions + QueryNetwork>(
 	validator: &PlutusScript,
 	policy: &PlutusScript,
 	d_parameter: &DParameter,
 	ctx: &TransactionContext,
-	mint_witness_ex_units: ExUnits,
+	mut ex_units: ScriptExUnits,
+	genesis_utxo: UtxoId,
+	client: &C,
 ) -> Result<Transaction, JsError> {
 	let mut tx_builder = TransactionBuilder::new(&get_builder_config(ctx)?);
 	// The essence of transaction: mint token and set output with it
-	tx_builder.add_mint_one_script_token(policy, mint_witness_ex_units)?;
+	tx_builder.add_mint_one_script_token(policy, ex_units.pop_mint_ex_units())?;
 	tx_builder.add_output_with_one_script_token(
 		validator,
 		policy,
 		&d_parameter_to_plutus_data(d_parameter),
 		ctx,
 	)?;
 
+	let gov_utxo = crate::init_governance::get_governance_utxo(genesis_utxo, client).await.map_err(|e| JsError::from_str(e.to_string().as_str()))?;
+	let gov_policy = match gov_utxo.script {
+		Some(Plutus(ps)) => PlutusScript::from_cbor(&ps.cbor, LanguageKind::PlutusV2),
+		_ => return Err(JsError::from_str("Governance UTXO script is not PlutusScript")),
+	};
+
+	let gov_tx_input = TransactionInput::new(&TransactionHash::from_bytes(gov_utxo.transaction.id.into())?, gov_utxo.index.into());
+    tx_builder.add_mint_one_script_token_using_reference_script(&ScriptHash::from_bytes(gov_policy.script_hash().to_vec())?, &gov_tx_input, gov_policy.bytes.len(), ex_units.pop_mint_ex_units())?;
+
+
+	let tx_hash = TransactionHash::from_bytes(gov_utxo.transaction.id.into())?;
+	let gov_tx_input = TransactionInput::new(&tx_hash, gov_utxo.index.into());
+	tx_builder.add_script_reference_input(&gov_tx_input, gov_policy.bytes.len());
+	tx_builder.add_required_signer(&ctx.payment_key_hash());
 	tx_builder.balance_update_and_build(ctx)
 }
 
-fn update_d_param_tx(
+async fn update_d_param_tx<C : QueryLedgerState + Transactions + QueryNetwork>(
 	validator: &PlutusScript,
 	policy: &PlutusScript,
 	d_parameter: &DParameter,
 	script_utxo: &OgmiosUtxo,
 	ctx: &TransactionContext,
-	validator_redeemer_ex_units: ExUnits,
+	mut ex_units: ScriptExUnits,
+	genesis_utxo: UtxoId,
+	client: &C,
```

Comment:
Same as for `mint_d_param_token_tx` - please keep it clean and make dirty stuff in some function on the higher level.